usr/man/_vax.d/man8/screend.8
.TH screend 8
.SH Name
screend \- Internet (IP) gateway screening daemon
.SH Syntax
.B /usr/etc/screend
[
.B \-d
] [
.B \-c
] [
.B \-l
] [
.B \-f
.I configfile
] [
.B \-L
.I logfile
] [
.B \-s
] [
.B \-r
]
.SH Description
.NXR "screend server daemon"
.NXR "Internet gateway screening daemon"
The
.PN screend
daemon is used in conjunction with the
gateway screen facility 
to decide which IP packets should be forwarded when the 
system is acting as an IP gateway.  Only the super-user may run
this program.  
.PP
Before using 
.PN screend
you must add the following line to your system configuration file:
.EX
pseudo-device      gwscreen
.EE
After editing the system configuration file you must rebuild
the kernel, and reboot the system.  For information on
rebuilding the kernel see the
\fIGuide to Configuration File Maintenance\fP
and the 
.MS doconfig 8
reference page.
.PP
When
.PN screend
is started, it reads the configuration file specified
(\fIconfigfile\fR) and then processes packets according to the
instructions in that file.
.PP
The kernel screening facility must be enabled using
.MS screenmode 8 
before
.PN screend
has any effect.
(When screening is disabled, packets are forwarded by the kernel
according to its usual procedures.)
.PP
It is possible to run more than one copy of
.PN screend
at a time, although it is not recommended.
You might do this, however, when the configuration file is changed.
Because the
configuration file is read only at program startup, you must restart
.PN screend
when the file is changed.  To avoid any service interruption, you
should start a new instance of
.PN screend 
before killing the old one.
.SH Options
.IP \-d 5
Prints large amounts of debugging information.  This is not
meant for normal use.
.IP \-c
Checks the syntax of the configuration file but does not actually
process any packets.
.IP \-l
Turns on logging for all packets (normally, packets are logged only
if requested in the configuration file).
.IP "\-f \fIconfigfile\fP"
Specifies the configuration file.  If not specified, the default,
.PN /etc/screend.conf , 
is used.
.IP "\-L \fIlogfile\fP"
Specifies that logging records should be appended to the given file.
There is no default logging file.
This may be specified simultaneously with 
.PN \-s , 
in which case each logging record is stored by both means.
.IP \-s
Specifies that logging records should be logged using
.MS syslog 8 .
.IP \-r
Specifies that logging records should include the rule number
of the configuration file rule responsible for the action logged.
This is useful for debugging configuration file problems.
See the section in this reference page on Rule Numbers.
.SH Configuration File
This is an informal guide to the grammar of the
.PN screend
configuration file.  It is intended for readers who
are familiar with the basic concepts of the IP protocol family,
including the distinction between the terms ``network'' and ``subnet.''
.PP
Lexical structure:
.RS
.IP Comments 
Can either be C-style comments, delimited by
/* and */
or csh-style comments begun with a number sign (\&#) and 
terminated by the end of a line.  Comments do not nest.
.IP Case 
Significant in reserved words (all are lower-case).
This is actually a benefit, because if a host name happens
to conflict with a reserved word, you can use the
host name in upper-case.
.IP "Host names"
Must begin with a letter but may contain
digits, minus signs (\-), 
dots ( . ), and underscores (\&_ ).  
The same is true of network,
subnet, and netmask names.  Hosts can also be identified by 
their IP address, in dotted quad notation (for example, ``128.45.40.15'').
.IP Numbers 
May be in decimal or in hex (0x0 notation).
Octal notation is not allowed.  Decimal notation is the preferred method.
.IP "Protocol names" 
Specified as they are found in 
.PN /etc/protocols .
These can also be given as numbers.
.IP "Port names" 
For TCP or UDP, specified as they are in
.PN /etc/services .
These can also be given as numbers (host byte order).
.IP "ICMP type codes" 
Must be chosen from the following list, or given as
numbers:
.IP
.\"	###################################################
.\" Documentation writer: PLEASE do not change this back to a "tbl"
.\" table, or it won't work with the online "man" command!
.\"	###################################################
.\".TS
.ta +\w'addressmaskrequest  'u +\w'addressmaskreply  'u
echo	echoreply	sourcequench
redirect	unreachable	timeexceeded
parameterproblem	timestamp	timestampreply
informationrequest	informationreply	
addressmaskrequest	addressmaskreply	
.\".TE
.IP "White space" 
All white space is the same (including newlines).
.RE
.PP
General syntax rules:
.IP 
The configuration file consists of specifications terminated by
semicolons.
.IP 
There are three kinds of specifications:
.RS
.IP "default-action specification" 
There should only be one of these (the last
one is the one that counts); it specifies what action to take
if no action specification matches a packet.
.IP "subnet mask specifications" 
Specifies the subnet mask used
for a given network.
.IP "action specifications" 
Specifies a class of packets and
the action to take when such a packet is received.
.RE
.IP
Specifications can appear in any order, but the evaluation
order of action specifications is the order in which
they appear in the file.
.PP
In BNF, this is:
.EX 0
\fI<configuration-file>\fR ::= \
{ \fI<specification>\fR | \fI<configuration-file>\fR \fI<specification>\fR }
\fI<specification>\fR ::= \
{ \fI<default-action>\fR | \fI<subnet-spec>\fR | \fI<action-spec>\fR }
.EE
.PP
The syntax for a default-action specification is:
.EX 0
\fI<default-action>\fR ::= \
\fBdefault\fR {\fBaccept\fR | \fBreject\fR} [\fBnotify\fR] [\fBlog\fR] \fB;\fR
.EE
Note that
.PN "default accept notify;" 
is not legal.
If not specified, the default-action is 
.PN reject .
.PP
The syntax for subnet mask specifications is:
.EX 0
\fI<subnet-spec>\fR ::= \
\fBfor\fR \fI<network>\fR \fBnetmask is\fR \fI<maskval>\fR \fB;\fR
.EE
The \fI<network>\fR is either a network name or a dotted-quad address,
such as ``36.0.0.0''.  The number ``36'' is 
not a reasonable value.
\fI<Maskval>\fR is either a name (treated as a hostname) or a dotted-quad
address, such as ``255.255.255.0'' (bits are \fIon\fR for the network
and subnet fields.)
.PP
The syntax for action specifications is:
.EX 0
\fI<action-spec>\fR ::= \
\fBfrom\fR \fI<object>\fR \fBto\fR \fI<object>\fR \
{\fBaccept\fR | \fBreject\fR} [\fBnotify\fR] [\fBlog\fR] \fB;\fR
.EE
Such a specification
says that packets flowing this way between this pair of
objects
(defined below) should either be accepted or rejected.  If 
.PN notify
is specified, when a packet is rejected an ICMP error message is
returned to the source.  If 
.PN log 
is specified, this packet and its disposition are logged.
.PP
Conceptually, for each packet the action specifications are
searched in the order they appear in the configuration file, until
one matches.  The specified action is then performed.  If no specification
matches, the default action is performed.
.PP
To simplify the configuration file, the following syntax 
may be used to indicate that the same action should be performed
on packets flowing in either direction between the specified pair
of objects:
.EX 0
\fI<action-spec>\fR ::= \
\fBbetween\fR \fI<object>\fR \fBand\fR \fI<object>\fR \
{\fBaccept\fR | \fBreject\fR} [\fBnotify\fR] [\fBlog\fR] \fB;\fR
.EE
Note that this has the 
same effect as specifying the two unidirectional rules, with the
forward direction listed first.
.PP
An object is a specification of the source or destination
of a packet.
The syntax for object specifications is 
somewhat complex, since certain fields are optional:
.EX 0
\fI<object>\fR ::= \
{ \fI<address-spec>\fR | \fI<port-spec>\fR | \
\fI<address-spec>\fR \fI<port-spec>\fR }
.EE
If the \fI<address-spec>\fR is not given, any host will match.
If the \fI<port-spec>\fR is not given, any protocol and port will match.
.EX 0
\fI<address-spec>\fR ::= \
{ \fI<net-spec>\fR | \fI<subnet-spec>\fR | \fI<host-spec>\fR | \fBany\fR }

\fI<net-spec>\fR ::= \
{ \fBnet\fR \fI<name-or-addr>\fR | \fBnet\-not\fR \fI<name-or-addr>\fR }
\fI<subnet-spec>\fR ::= \
{ \fBsubnet\fR \fI<name-or-addr>\fR | \fBsubnet\-not\fR \fI<name-or-addr>\fR }
\fI<host-spec>\fR ::= \
{ \fBhost\fR \fI<name-or-addr>\fR | \fBhost\-not\fR \fI<name-or-addr>\fR }
.EE
The 
.PN \-not 
convention means that the object specification matches
if the specified field does
.I not
have the specified
value.  In the following example, packets not from
nic.ddn.mil are dropped.
.EX
from host\-not nic.ddn.mil to host any reject;
.EE
The ``subnet'' and ``subnet\-not''
forms match against the
entire address under the subnet mask (for example, if the
netmask for net 36 is ``255.255.0.0'', then ``subnet
36.8.0.0'' matches a packet address of 36.8.0.1).
.PP
.EX 0		
\fI<name-or-addr>\fR ::= \
{ \fI<name>\fR | \fI<dotted-quad>\fR | \fBany\fR }

\fI<port-spec>\fR ::= { \fBproto\fR \fI<proto-name-or-number>\fR
.RS 
| \fBicmp type\fR \fI<type-name-or-number>\fR \
| \fBicmp type\-not\fR \fI<type-name-or-number>\fR
| \fBtcp port\fR \fI<port-name-or-number>\fR \
| \fBtcp port\-not\fR \fI<port-name-or-number>\fR
| \fBudp port\fR \fI<port-name-or-number>\fR \
| \fBudp port\-not\fR \fI<port-name-or-number>\fR }
.RE

\fI<proto-name-or-number>\fR ::= { \fI<name>\fR | \fI<number>\fR }
\fI<type-name-or-number>\fR ::= \
{ \fI<name>\fR | \fI<number>\fR | \fBany\fR |  \fBinfotype\fR }
\fI<port-name-or-number>\fR ::= \
{ \fI<name>\fR | \fI<number>\fR | \fBany\fR | \fBreserved\fR  | \fBxserver\fR }
.EE
``Reserved'' ports are those reserved by 4.2BSD Unix for
privileged processes.
``Xserver'' ports are those used by X11 window system servers.
``Infotype'' ICMP packets are those that are
purely informational: echo, timestamp, information, and addressmask
requests, and the corresponding replies.
.SH Restrictions
IP gateways are allowed to
fragment IP datagrams if they are too large to be forwarded in one piece.
Only the first fragment of a datagram carries enough information
to make certain kinds of accept/reject decisions.
The 
.PN screend
daemon can only handle fragments if it sees the first fragment of a
datagram before it sees any subsequent fragments.
Also, only a limited rate of fragmented packet arrival can be
accommodated by the program (fragmentation is, in general, a bad idea).
Finally, if more than one instance of
.PN screend
is running, most likely this will result in significant loss of
fragments.
.PP
The current implementation does not forward packets that contain
IP header options.  This is because several of these options can
be used to subvert checks based on the IP header destination address.
.PP
If a host 
.I name 
given in an object specification has more than
one IP address associated with it,
.PN screend
does not understand that all these addresses should be checked.
Only the first (primary) address of the host is used.  This may
lead to erroneous operation in some cases (possibly including a
security hole), so a warning is printed if the configuration file
contains such names.  (Note that you probably will not see this warning
if
.PN screend
is only started in 
.PN /etc/rc .)
.SH Examples
This following is an example of the syntax; it is not intended to
be used in an actual installation:
.EX
# Example configuration file
default reject;

for 36.0.0.0 netmask is 255.255.0.0;

from subnet 36.8.0.0 to net milnet reject notify;
from host nic.ddn.mil to host any accept;
from host any to net arpanet tcp port telnet accept;
from host any to host any icmp type redirect reject log;
from host any to subnet 36.10.0.0 tcp port-not reserved reject;
.EE
.SH Rule Numbers
If the 
.PN \-r 
option is given, log records contain a notation
of the rule number responsible for the action being logged.
A rule is a ``from ... to ...'' specification
in the configuration file; rules are numbered in order starting with
zero.
Note that ``between ... and ...'' specifications expand to two
``from ... to ... '' rules, each numbered individually.
The default action, whether explicitly stated or not, is
not numbered; it is referred to distinctively in the log.
.SH Diagnostics
During argument processing and configuration file parsing,
various diagnostics may be issued.  During normal operation,
only serious internal inconsistencies result in diagnostics.
(See the Restrictions section about warning messages in some
borderline cases.)
Except in debug mode (
.PN \-d ), 
most diagnostics are logged using
.MS syslog 8 .
.PP
Once an hour, a statistics report is made using
.MS syslog 8
that shows the number of packets processed since the program was
started, the hit rate of an internal cache buffer, and the number
of packets dropped because they arrived too rapidly.
.SH Files
.TP 25
.PN /etc/screend.conf       
default configuration file
.SH See Also
screen(2), screenmode(8), screenstat(8)


usr/etc/screend.d/screend.c
#ifndef lint
static char *sccsid = "@(#)screend.c	4.2	(ULTRIX)	10/12/90";
#endif
/*
 * screend.c
 *
 * IP gateway screening daemon
 *
 * Usage:
 *	screend [-d] [-c] [-l] [-f filename] [-s] [-r] [-L logfile]
 *
 * Modification History:
 *
 * 21 December 1988	Jeffrey Mogul/DECWRL
 *	Created.
 */
/*
 *               Copyright 1989, 1990 Digital Equipment Corporation
 *                          All Rights Reserved
 * 
 * 
 * Permission to use, copy, and modify this software and its documentation
 * is hereby granted only under the following terms and conditions.  Both
 * the above copyright notice and this permission notice must appear in
 * all copies of the software, derivative works or modified versions, and
 * any portions threof, and both notices must appear in supporting
 * documentation.
 * 
 * Users of this software agree to the terms and conditions set forth herein,
 * and hereby grant back to Digital a non-exclusive, unrestricted, royalty-free
 * right and license under any changes, enhancements or extensions made to the
 * core functions of the software, including but not limited to those affording
 * compatibility with other hardware or software environments, but excluding
 * applications which incorporate this software.  Users further agree to use
 * their best efforts to return to Digital any such changes, enhancements or
 * extensions that they make and inform Digital of noteworthy uses of this
 * software.  Correspondence should be provided to Digital at:
 * 
 *                       Director of Licensing
 *                       Western Research Laboratory
 *                       Digital Equipment Corporation
 *                       100 Hamilton Avenue
 *                       Palo Alto, California  94301  
 * 
 * Comments and bug reports may also be sent using electronic mail to:
 * 			screend-reports@decwrl.dec.com
 * 
 * 	>> This software may NOT be distributed to third parties. <<
 *   
 * THE SOFTWARE IS PROVIDED "AS IS" AND DIGITAL EQUIPMENT CORP. DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS.   IN NO EVENT SHALL DIGITAL
 * EQUIPMENT CORPORATION BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
 * USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/file.h>
#include <sys/errno.h>

#include <net/if.h>
#include <netinet/in.h>
#include "screentab.h"
#include <net/gw_screen.h>

#include <stdio.h>
#include <syslog.h>
#include <signal.h>

#define	DEFCONFIG "/etc/screend.conf"

#define	REPORT_INTERVAL	(60*60)		/* seconds between stat reports */

char *sourcefile = DEFCONFIG;

int debug = 0;
int semantic_errors = 0;
int logall = 0;
int checkonly = 0;
int use_syslog = 0;
int log_rule = 0;
int compress_log = 1;	/* do we even need a flag for this one? */
FILE *log_file = NULL;
char *logfilename = NULL;

extern int LastMatchRule;
extern int LastCacheHit;

int	TotalCacheHits;
int	TotalCacheMisses;

int	screen_sock;

void Restart();
void Reporter();

extern int errno;

main(argc, argv)
int argc;
char **argv;
{
	struct itimerval itv;

	while (argc > 1) {
	    if (argv[1][0] != '-') {
		Usage();
		exit(1);
	    }
	    switch (argv[1][1]) {
		case 'c':
		    checkonly++;
		    break;
		case 'd':
		    debug++;
		    break;
		case 'l':
		    logall++;
		    break;
		case 'L':
		    argc--;
		    argv++;
		    if (argc < 2) {
			fprintf(stderr, "-L: no filename argument\n");
			Usage();
			exit(1);
		    }
		    logfilename = argv[1];
		    break;
		case 'f':
		    argc--;
		    argv++;
		    if (argc < 2) {
			fprintf(stderr, "-f: no filename argument\n");
			Usage();
			exit(1);
		    }
		    sourcefile = argv[1];
		    break;
		case 's':
		    use_syslog = 1;
		    break;
		case 'r':
		    log_rule = 1;
		    break;
		default:
		    fprintf(stderr, "%s: unknown flag\n", argv[1]);
		    Usage();
		    exit(1);
		    break;
	    }
	    argc--;
	    argv++;
	}

	if (freopen(sourcefile, "r", stdin) == NULL) {
	    perror(sourcefile);
	    exit(1);
	}

	InitTables();
	InitFragCache();
	if (yyparse() || semantic_errors) {
	    fprintf(stderr, "%s: not correct\n", sourcefile);
	    exit(1);
	}
	fclose(stdin);	/* don't need config file any more */

	if (debug) {
	    DumpNetMaskTable();
	    DumpActionTable();
	}
	
	if (checkonly) {
	    exit(0);
	}

	if (debug == 0) {
	    int i;

	    if (fork()) {
		exit(0);
	    }

	    i = getdtablesize();
	    while (--i >= 0)
		(void)close(i);

	    (void) open("/dev/null", O_RDONLY);
	    (void) dup2(0, 1);
	    (void) dup2(0, 2);
	    i = open("/dev/tty", O_RDWR);
	    if (i >= 0) {
		(void) ioctl(i, TIOCNOTTY, NULL);
		(void) close(i);
	    }
	}

	if (use_syslog) {
	    if (openlog("screend", LOG_PID|LOG_TIME) < 0) {
		perror("syslog");
	    }
	    else {
		LogSyslog("started");
	    }
	}

	if (logfilename) {
	    log_file = fopen(logfilename, "a");
	    if (log_file == NULL) {
		perror(logfilename);
		exit(1);
	    }
	    LogToFileTS(log_file, "START");
	}

	screen_sock = socket(AF_INET, SOCK_DGRAM, 0);
	if (screen_sock < 0) {
		Myperror("screend: socket");
		exit(1);
	}

	(void)signal(SIGHUP, Restart);
	(void)signal(SIGALRM, Reporter);

	/*
	 * Set up a periodic timer for doing statistics reports;
	 * do the first report "real soon" so that we have a
	 * baseline.
	 */
	itv.it_interval.tv_sec = REPORT_INTERVAL;
	itv.it_interval.tv_usec = 0;
	itv.it_value.tv_sec = 1;
	itv.it_value.tv_usec = 0;
	
	(void) setitimer(ITIMER_REAL, &itv, (struct itimerval *) 0);

	InitReporter(screen_sock);

	Daemon(screen_sock);
}

Usage()
{
	fprintf(stderr,
"Usage: screend [-d] [-c] [-l] [-f filename] [-s] [-r] [-L logfile]\n");
}

/* Entered upon SIGHUP */
void Restart()
{
	Reporter();
	if (use_syslog)
	    syslog(LOG_INFO, "restarting due to SIGHUP", 0);
	else
	    syslog(LOG_INFO, "screend: restarting due to SIGHUP", 0);
	ReOpenLog(1);
	InitReporter(screen_sock);
	if (use_syslog)
	    syslog(LOG_INFO, "restart complete", 0);
	else
	    syslog(LOG_INFO, "screend: restart complete", 0);
}


struct	screen_data scdata;

Daemon(s)
register int s;
{
	static struct annotated_hdrs ah;
	register int actcode;

	scdata.sd_xid = 0;	/* make sure things start "clean" */
	scdata.sd_family = AF_INET;	/* make sure we get only IPs */
	scdata.sd_action = SCREEN_DROP;

	while (1) {
	    if (debug) {
		PrintScreenData(&scdata);
	    }
	    if (ioctl(s, SIOCSCREEN, (caddr_t)&scdata) < 0) {
		if (errno != EINTR) {
		    Myperror("ioctl (SIOCSCREEN)");
		    exit(1);
		}
	    }
	    scdata.sd_action = SCREEN_DROP;	/* default is DROP */

	    /* unpack this packet */
	    bzero(&ah, sizeof(ah));
	    if (UnpackIP(scdata.sd_data, scdata.sd_dlen, &(ah.hdrs)) == 0) {
		/* cannot parse this packet */
		continue;
	    }

	    if (debug) {
		PrintAnnotatedHdrs(&ah);
		printf("\n");
	    }
	    
	    /* Now see what we have */
	    actcode = ScreenIt(&ah);

	    if (LastCacheHit)
		TotalCacheHits++;
	    else
		TotalCacheMisses++;

	    if (actcode & ASACTION_ACCEPT)
		scdata.sd_action = SCREEN_ACCEPT;
	    else if (actcode & ASACTION_NOTIFY)
		scdata.sd_action |= SCREEN_NOTIFY;
	    if (logall || (actcode & ASACTION_LOG))
		LogIt(&scdata, actcode, &(ah.hdrs));
	}
}

PrintScreenData(sdp)
register struct screen_data *sdp;
{
	printf("(%d.%06d) af %d count %d dlen %d xid %x action %x",
		sdp->sd_arrival.tv_sec, sdp->sd_arrival.tv_usec,
		sdp->sd_family,
		sdp->sd_count, sdp->sd_dlen, sdp->sd_xid, sdp->sd_action);
	if (sdp->sd_action & SCREEN_ACCEPT)
		printf(" ACCEPT");
	else
		printf(" REJECT");
	if (sdp->sd_action & SCREEN_NOTIFY)
		printf(" NOTIFY");
	printf("\n");

/*	PrintIPHeader(&(sdp->sd_arrival), sdp->sd_data, sdp->sd_dlen); */
}

int origdrops;

InitReporter(s)
int s;
{
	struct screen_stats sstats;
	
	if (ioctl(screen_sock, SIOCSCREENSTATS, (caddr_t)&sstats) < 0) {
	    Myperror("ioctl (SIOCSCREENSTATS)");
	    origdrops = 0;
	    return;
	}
	
	origdrops = sstats.ss_nobuffer + sstats.ss_badsync + sstats.ss_stale;
	TotalCacheHits = 0;
	TotalCacheMisses = 0;
}

void Reporter()
{
	struct screen_stats sstats;
	
	if (ioctl(screen_sock, SIOCSCREENSTATS, (caddr_t)&sstats) < 0) {
	    Myperror("ioctl (SIOCSCREENSTATS)");
	    return;
	}
	
	LogStats(TotalCacheHits, TotalCacheMisses, origdrops, &sstats);

	/*
	 * Once an hour, close and open log file (if in use) so that
	 * renames of the log file will have some effect.
	 */
	ReOpenLog(0);
}
